README
                              X3 Readme File

Description:
----------
X3 is a complete set of services for Nefarious IRCu P10 protocol
networks based on srvx irc services.. It aims to be highly efficient,
flexible, and extensible. Features include advanced proxy detection
and all the standard services users expect e.g. ChanServ, NickServ,
OpServ, HelpServ, and Global.

X3 is open source software; the source is available for download
under the terms and conditions of the GNU General Public License,
commonly known as the GNU GPL.

The official X3 web site is available at http://evilnet.sourceforge.net

Installation:
----------
Please read the INSTALL file for installation instructions.

Support:
----------
There are several conditions under which we WILL NOT support your use
of these services. (The license may permit such uses, but we do not
wish to encourage them.)

Specifically:

 - Use under Windows. X3 can work to some extent under the Cygwin
 (http://www.cygwin.com/) environment, but Windows as a whole departs
 significantly from standard APIs and is difficult to support.

 - Alteration of credits. If you change the credits in your version of
 the code so as to obscure the original authorship of the program, we
 will not support you.  If you want to use our code -- especially if
 you want us to answer your questions about it -- do the decent thing
 and preserve the credits that are in the code we distribute.

In addition, please read the documentation included in the X3
distribution before you ask us for support. The answers to many common
questions related to installation are in the INSTALL file, for
example. It saves both our time and your time, so please read the
documentation.

Contacts:
----------
The AUTHORS file contains contact information for the X3 Development
Team, aswell as contact information for the original srvx Development
Team.

t


ChangeLog
/***********************************************************************
X3 ChangeLog

2013-02-18  Alex Schumann  <rubin@afternet.org>
	* all: Converted from mercurial to git. (testing email commit messages again)
	* configure.in: Roll version to 1.8.1
	* src/version.c.SH: replace hg version string with git one

2013-01-27  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/nickserv.c: Fix for crash bug when using calling *NickServ.oset
	with an account name where nobody is online using that account and no
	fakehost is set.

2012-04-29  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/x3ldap.c: Replaced ldap_get_values() with ldap_get_values_len().

2012-01-26  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/dict-splay.c: Fix for possible crash bugs in dict_splay()
	and dict_remove2();

	* src/tools.c: Fix for crash bug in irccasecmp();

2012-01-24  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/proto-p10.c: Fixed irc_topic() to honour server/hidden_host_type
	and associated config settings.

2012-01-16  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/chanserv.c: Fix for USER AUTOOP option being ignored during
	ChanServ.resync.

2012-01-13  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/opserv.help: Fixed typo in help for TRACE CRITERIA.

2011-11-16  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/nickserv.c: Added a new line to NickServ.accountinfo to show
	new email address for email change cookies.

2011-10-12  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/chanserv.help: Fixed typo in staff register help.

2011-09-29  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/chanserv.help: Fixed typo in durations help.

2011-07-18  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/proto-p10.c: Fix for "PARSE ERROR" on "Ghost 5 Numeric Collided"
	kill message.

2011-06-29  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/nickserv.c: Fixed a bug when reading services/nickserv/hard_maxlogins
	from x3.conf

2011-01-25  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/proto-p10.c: Fix for auto-oper bug where PRIVS were not
	being applied on Nefarious branch versions (or in short reverted
	revision 697:e6fe90cab8d0)

2010-12-06  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/nickserv.c: Fixed bug where NickServ.oregister incorrectly
	logged the password.

2010-10-15  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/proto-p10.c: Fix for P10 BURST message parsing where
	extended bans were causing premature change from ban mask to
	except mask parsing.

2010-09-27  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/chanserv.c: Fix for maxsetinfo now allowing you to set it
	to the maximum value as specified in x3.conf

2010-09-24  Matthew Beeching  <jobe@mdbnet.co.uk>

        * src/proto-p10.c: Fix for crash bug during burst where X3 sends
        a B (BURST) message with a channel limit in the modes string.

2010-09-24  Alex Schumann <rubin@afternet.org>

	* src/chanserv.c: dont let users set extended bans with
	.ban/lamer/etc, and dont remove extended bans with bantimeout

2010-09-19  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/chanserv.c: Added ignore_csuspend to *ChanServ.csuspend
	and *ChanServ.cunsuspend as a default flag.

2010-09-17  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/chanserv.c: Updated ChanServ.staff, ChanServ.ircops and
	ChanServ.helpers to hide users with user mode +H (hide oper).

2010-09-08  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/proto-p10.c: Replaced some horrible code for formatting a
	channel mode string and fix for a bug where chan mode +A was
	being ignored.

2010-08-29  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/chanserv.c: Changed wheel of misfortune nick change pseudo
	user to use a hard coded host name to prevent accidentally giving
	out a users real host.

2010-08-19  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/proto-p10.c: Changed TOK_PRIVS to "PR" to match current ircu
	distributions.

2010-07-31  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/proto-common.c: Added hg revision information to VERSION
	reply to match version strings everywhere else.

2010-07-30  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/Makefile.in: Added version.c generation at compile time

	* src/modcmd.c: Updated to use cvs_version from version.c

	* src/version.c.SH: Added version.c.SH to generate version.c

	* src/version.h: Added version.h

2010-07-07  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/hash.h: Added support for user modes +D, +R, +W, +H, +L
	and +q.

	* src/opserv.c: Added support for user modes +D, +R, +W, +H, +L
	and +q.

	* src/proto-p10.c: Added support for user modes +D, +R, +W, +H,
	+L and +q.

2010-07-07  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/hash.h: Added support for user modes +z and +a and fixed
	some comments.

	* src/opserv.c: Added support for user modes +z and +a and
	added support to TRACE/ADDALERT CHECKRESTRICTIONS criteria for
	channel mode +a and +Z.

	* src/proto-p10.c: Added support for user modes +z and +a.

2010-07-07  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/opserv.c: Added support to TRACE/ADDALERT CHECKRESTRICTIONS
	criteria for channel mode +O and +r.

2010-07-03  Matthew Beeching  <jobe@mdbnet.co.uk>

	* src/chanserv.c: Removed extra 'is' from CSMSG_SMURF_TARGET.

2010-06-06  Matthew Beeching  <jobe@mdbnet.co.uk>


src/md5.c
/*
 *  RFC 1321 compliant MD5 implementation
 *
 *  Copyright (C) 2001-2003  Christophe Devine
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <string.h>

#include "common.h"
#include "md5.h"

#define OLDPASSFUNC
#ifdef OLDPASSFUNC
/* ----------------------------------------------------- */
/*         DELETE THIS SOME DAY - OLD STUFF HERE         */
/* ----------------------------------------------------- */

/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */

/* Constants for MD5Transform routine.
 */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static void MD5Transform PROTO_LIST ((UINT4 [4], unsigned char [64]));
static void Encode PROTO_LIST
  ((unsigned char *, UINT4 *, unsigned int));
static void Decode PROTO_LIST
  ((UINT4 *, unsigned char *, unsigned int));

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits. */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

/* MD5 initialization. Begins an MD5 operation, writing a new context. */
void MD5Init (context)
MD5_CTX *context;                                        /* context */
{
  context->count[0] = context->count[1] = 0;
  /* Load magic initialization constants. */
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
void MD5Update (context, input, inputLen)
MD5_CTX *context;                                        /* context */
unsigned char *input;                                /* input block */
unsigned int inputLen;                     /* length of input block */
{
  unsigned int i, idx, partLen;

  /* Compute number of bytes mod 64 */
  idx = (unsigned int)((context->count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((context->count[0] += ((UINT4)inputLen << 3))
      < ((UINT4)inputLen << 3))
      context->count[1]++;
  context->count[1] += ((UINT4)inputLen >> 29);

  partLen = 64 - idx;

  /* Transform as many times as possible. */
  if (inputLen >= partLen) {
      memcpy((POINTER)&context->buffer[idx], (POINTER)input, partLen);
      MD5Transform (context->state, context->buffer);

      for (i = partLen; i + 63 < inputLen; i += 64)
	  MD5Transform (context->state, &input[i]);

      idx = 0;
  }
  else
    i = 0;

  /* Buffer remaining input */
  memcpy((POINTER)&context->buffer[idx], (POINTER)&input[i],
  inputLen-i);
}

/* MD5 finalization. Ends an MD5 message-digest operation, writing the
  the message digest and zeroizing the context.
 */
void MD5Final (digest, context)
unsigned char digest[16];                         /* message digest */
MD5_CTX *context;                                       /* context */
{
  unsigned char bits[8];
  unsigned int idx, padLen;

  /* Save number of bits */
  Encode (bits, context->count, 8);

  /* Pad out to 56 mod 64. */
  idx = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (idx < 56) ? (56 - idx) : (120 - idx);
  MD5Update (context, PADDING, padLen);

  /* Append length (before padding) */
  MD5Update (context, bits, 8);

  /* Store state in digest */
  Encode (digest, context->state, 16);

  /* Zeroize sensitive information. */
  memset ((POINTER)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block. */
static void MD5Transform (state, block)
UINT4 state[4];
unsigned char block[64];
{
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode (x, block, 64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  /* Zeroize sensitive information. */
  memset ((POINTER)x, 0, sizeof (x));
}

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
static void Encode (output, input, len)
unsigned char *output;
UINT4 *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
    output[j] = (unsigned char)(input[i] & 0xff);
    output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
    output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
    output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4.
 */
static void Decode (output, input, len)
UINT4 *output;
unsigned char *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
    output[i] = ((UINT4)input[j]) | (((UINT4)input[j+1]) << 8) |
      (((UINT4)input[j+2]) << 16) | (((UINT4)input[j+3]) << 24);
}

static const char *
cryptpass_real(const char *pass, char *buffer, int seed)
{
    static const char hex_digits[] = "0123456789ABCDEF";
    MD5_CTX ctx;
    unsigned int len, i, j;
    unsigned char temp[64], buff[16];

    /* first: generate seed */
    if (seed) {
        for (i=j=0; j<4; j++) {
            temp[i++] = hex_digits[(seed >> (28 - 8*j)) & 15];
            temp[i++] = hex_digits[(seed >> (24 - 8*j)) & 15];
        }
        /* fill temp buffer with password, 1 and then 0's */
        j = strlen(pass);
        if ((sizeof(temp) - i) < j) j = sizeof(temp) - i;
        memcpy(temp+i, pass, j);
        i += j;
        if (i < sizeof(temp)) temp[i++] = '1';
        while (i < sizeof(temp)) temp[i++] = '0';
    } else {
        i = 0;
        /* next: duplicate password to fill temp buffer */
        len = strlen(pass);
        for (j=0; i<sizeof(temp); i++) {
            temp[i] = pass[j];
            if (++j == len) j = 0;
        }
    }
    /* next: compute MD5 digest of repeated password */
    MD5Init(&ctx);
    MD5Update(&ctx, temp, sizeof(temp));
    MD5Final(buff, &ctx);
    /* finally: write digest to output buffer */
    if (seed) {
        buffer[0] = '$';
        for (i=0,j=1; i<4; i++) {
            buffer[j++] = hex_digits[(seed >> (28 - 8*i)) & 15];
            buffer[j++] = hex_digits[(seed >> (24 - 8*i)) & 15];
        }
    } else {
        j = 0;
    }
    for (i=0; i<sizeof(buff); i++) {
	buffer[j++] = hex_digits[buff[i] >> 4];
	buffer[j++] = hex_digits[buff[i] & 15];
    }
    buffer[j] = 0;
    return buffer;
}

/* ----------------------------------------------------- */
/*           DELETE ABOVE SOME DAY -- OLD STUFF          */
/* ----------------------------------------------------- */

#endif

#define GET_UINT32(n,b,i)                       \
{                                               \
    (n) = ( (uint32) (b)[(i)    ]       )       \
        | ( (uint32) (b)[(i) + 1] <<  8 )       \
        | ( (uint32) (b)[(i) + 2] << 16 )       \
        | ( (uint32) (b)[(i) + 3] << 24 );      \
}

#define PUT_UINT32(n,b,i)                       \
{                                               \
    (b)[(i)    ] = (uint8) ( (n)       );       \
    (b)[(i) + 1] = (uint8) ( (n) >>  8 );       \
    (b)[(i) + 2] = (uint8) ( (n) >> 16 );       \
    (b)[(i) + 3] = (uint8) ( (n) >> 24 );       \
}

void md5_starts( md5_context *ctx )
{
    ctx->total[0] = 0;
    ctx->total[1] = 0;

    ctx->state[0] = 0x67452301;
    ctx->state[1] = 0xEFCDAB89;
    ctx->state[2] = 0x98BADCFE;
    ctx->state[3] = 0x10325476;
}

void md5_process( md5_context *ctx, uint8 data[64] )
{
    uint32 X[16], A, B, C, D;

    GET_UINT32( X[0],  data,  0 );
    GET_UINT32( X[1],  data,  4 );
    GET_UINT32( X[2],  data,  8 );
    GET_UINT32( X[3],  data, 12 );
    GET_UINT32( X[4],  data, 16 );
    GET_UINT32( X[5],  data, 20 );
    GET_UINT32( X[6],  data, 24 );
    GET_UINT32( X[7],  data, 28 );
    GET_UINT32( X[8],  data, 32 );
    GET_UINT32( X[9],  data, 36 );
    GET_UINT32( X[10], data, 40 );
    GET_UINT32( X[11], data, 44 );
    GET_UINT32( X[12], data, 48 );
    GET_UINT32( X[13], data, 52 );
    GET_UINT32( X[14], data, 56 );
    GET_UINT32( X[15], data, 60 );

#define S(x,n) ((x << n) | ((x & 0xFFFFFFFF) >> (32 - n)))

#define P(a,b,c,d,k,s,t)                                \
{                                                       \
    a += F(b,c,d) + X[k] + t; a = S(a,s) + b;           \
}

    A = ctx->state[0];
    B = ctx->state[1];
    C = ctx->state[2];
    D = ctx->state[3];

#undef F
#define F(x,y,z) (z ^ (x & (y ^ z)))

    P( A, B, C, D,  0,  7, 0xD76AA478 );
    P( D, A, B, C,  1, 12, 0xE8C7B756 );
    P( C, D, A, B,  2, 17, 0x242070DB );
    P( B, C, D, A,  3, 22, 0xC1BDCEEE );
    P( A, B, C, D,  4,  7, 0xF57C0FAF );
    P( D, A, B, C,  5, 12, 0x4787C62A );
    P( C, D, A, B,  6, 17, 0xA8304613 );
    P( B, C, D, A,  7, 22, 0xFD469501 );
    P( A, B, C, D,  8,  7, 0x698098D8 );
    P( D, A, B, C,  9, 12, 0x8B44F7AF );
    P( C, D, A, B, 10, 17, 0xFFFF5BB1 );
    P( B, C, D, A, 11, 22, 0x895CD7BE );
    P( A, B, C, D, 12,  7, 0x6B901122 );
    P( D, A, B, C, 13, 12, 0xFD987193 );
    P( C, D, A, B, 14, 17, 0xA679438E );
    P( B, C, D, A, 15, 22, 0x49B40821 );

#undef F

#define F(x,y,z) (y ^ (z & (x ^ y)))

    P( A, B, C, D,  1,  5, 0xF61E2562 );
    P( D, A, B, C,  6,  9, 0xC040B340 );
    P( C, D, A, B, 11, 14, 0x265E5A51 );
    P( B, C, D, A,  0, 20, 0xE9B6C7AA );
    P( A, B, C, D,  5,  5, 0xD62F105D );
    P( D, A, B, C, 10,  9, 0x02441453 );
    P( C, D, A, B, 15, 14, 0xD8A1E681 );
    P( B, C, D, A,  4, 20, 0xE7D3FBC8 );
    P( A, B, C, D,  9,  5, 0x21E1CDE6 );
    P( D, A, B, C, 14,  9, 0xC33707D6 );
    P( C, D, A, B,  3, 14, 0xF4D50D87 );
    P( B, C, D, A,  8, 20, 0x455A14ED );
    P( A, B, C, D, 13,  5, 0xA9E3E905 );
    P( D, A, B, C,  2,  9, 0xFCEFA3F8 );
    P( C, D, A, B,  7, 14, 0x676F02D9 );
    P( B, C, D, A, 12, 20, 0x8D2A4C8A );

#undef F
    
#define F(x,y,z) (x ^ y ^ z)

    P( A, B, C, D,  5,  4, 0xFFFA3942 );
    P( D, A, B, C,  8, 11, 0x8771F681 );
    P( C, D, A, B, 11, 16, 0x6D9D6122 );
    P( B, C, D, A, 14, 23, 0xFDE5380C );
    P( A, B, C, D,  1,  4, 0xA4BEEA44 );
    P( D, A, B, C,  4, 11, 0x4BDECFA9 );
    P( C, D, A, B,  7, 16, 0xF6BB4B60 );
    P( B, C, D, A, 10, 23, 0xBEBFBC70 );
    P( A, B, C, D, 13,  4, 0x289B7EC6 );
    P( D, A, B, C,  0, 11, 0xEAA127FA );
    P( C, D, A, B,  3, 16, 0xD4EF3085 );
    P( B, C, D, A,  6, 23, 0x04881D05 );
    P( A, B, C, D,  9,  4, 0xD9D4D039 );
    P( D, A, B, C, 12, 11, 0xE6DB99E5 );
    P( C, D, A, B, 15, 16, 0x1FA27CF8 );
    P( B, C, D, A,  2, 23, 0xC4AC5665 );

#undef F

#define F(x,y,z) (y ^ (x | ~z))

    P( A, B, C, D,  0,  6, 0xF4292244 );
    P( D, A, B, C,  7, 10, 0x432AFF97 );
    P( C, D, A, B, 14, 15, 0xAB9423A7 );
    P( B, C, D, A,  5, 21, 0xFC93A039 );
    P( A, B, C, D, 12,  6, 0x655B59C3 );
    P( D, A, B, C,  3, 10, 0x8F0CCC92 );
    P( C, D, A, B, 10, 15, 0xFFEFF47D );
    P( B, C, D, A,  1, 21, 0x85845DD1 );
    P( A, B, C, D,  8,  6, 0x6FA87E4F );
    P( D, A, B, C, 15, 10, 0xFE2CE6E0 );
    P( C, D, A, B,  6, 15, 0xA3014314 );
    P( B, C, D, A, 13, 21, 0x4E0811A1 );
    P( A, B, C, D,  4,  6, 0xF7537E82 );
    P( D, A, B, C, 11, 10, 0xBD3AF235 );
    P( C, D, A, B,  2, 15, 0x2AD7D2BB );
    P( B, C, D, A,  9, 21, 0xEB86D391 );

#undef F

    ctx->state[0] += A;
    ctx->state[1] += B;
    ctx->state[2] += C;
    ctx->state[3] += D;
}

void md5_update( md5_context *ctx, uint8 *input, uint32 length )
{
    uint32 left, fill;

    if( ! length ) return;

    left = ctx->total[0] & 0x3F;
    fill = 64 - left;

    ctx->total[0] += length;
    ctx->total[0] &= 0xFFFFFFFF;

    if( ctx->total[0] < length )
        ctx->total[1]++;

    if( left && length >= fill )
    {
        memcpy( (void *) (ctx->buffer + left),
                (void *) input, fill );
        md5_process( ctx, ctx->buffer );
        length -= fill;
        input  += fill;
        left = 0;
    }

    while( length >= 64 )
    {
        md5_process( ctx, input );
        length -= 64;
        input  += 64;
    }

    if( length )
    {
        memcpy( (void *) (ctx->buffer + left),
                (void *) input, length );
    }
}

static uint8 md5_padding[64] =
{
 0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

void md5_finish( md5_context *ctx, uint8 digest[16] )
{
    uint32 last, padn;
    uint32 high, low;
    uint8 msglen[8];

    high = ( ctx->total[0] >> 29 )
         | ( ctx->total[1] <<  3 );
    low  = ( ctx->total[0] <<  3 );

    PUT_UINT32( low,  msglen, 0 );
    PUT_UINT32( high, msglen, 4 );

    last = ctx->total[0] & 0x3F;
    padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );

    md5_update( ctx, md5_padding, padn );
    md5_update( ctx, msglen, 8 );

    PUT_UINT32( ctx->state[0], digest,  0 );
    PUT_UINT32( ctx->state[1], digest,  4 );
    PUT_UINT32( ctx->state[2], digest,  8 );
    PUT_UINT32( ctx->state[3], digest, 12 );
}

#define TEST
#ifdef TEST

#include <stdlib.h>
#include <stdio.h>

/*
 * those are the standard RFC 1321 test vectors
 */

char* md5(const char* pass, char* output)
{
    md5_context ctx;
    int j;
    unsigned char md5sum[16];

    md5_starts( &ctx );
    md5_update( &ctx, (uint8 *) pass, strlen(pass));
    md5_finish( &ctx, md5sum );
    for(j = 0; j<16; j++)
    {
        sprintf(output + j * 2, "%02x", md5sum[j]);
    }
    /* printf("The hash of %s is %s\n\n", pass, output); */
    return 0;
}

const char *
cryptpass(const char *pass, char *buffer)
{
    return md5(pass, buffer);
}

int
checkpass(const char *pass, const char *crypted)
{
    char new_crypted[MD5_CRYPT_LENGTH], hseed[9];
    int seed;

    if (crypted[0] == '$') {
        /* new-style crypt, use "seed" after '$' */
        strncpy(hseed, crypted+1, 8);
        hseed[8] = 0;
        seed = strtoul(hseed, NULL, 16);
        cryptpass_real(pass, new_crypted, seed);
    } else {
        /* new "old-style" md5 crypt compatable with php, md5sum etc */
        md5(pass, new_crypted);
    }
    return !strcmp(crypted, new_crypted);
}



#endif
